tornado/test/websocket_test.py

```diff
@@ -92,6 +93,13 @@ def open(self, arg):
         self.write_message(arg)
 
 
+class CoroutineOnMessageHandler(TestWebSocketHandler):
+    @gen.coroutine
+    def on_message(self, message):
+        yield gen.sleep(0.1)
```

Comment:
I think what I'd do for testing is add a check that only one coroutine is sleeping at a time:

    if self.sleeping > 0:
        self.write_message('another coroutine is already sleeping')
    self.sleeping += 1
    yield gen.sleep(0.01)
    self.sleeping -= 1

And then in the tests themselves, write both messages before reading the first response.